core/parse-xml.cpp

```diff
@@ -340,9 +340,15 @@ static void temperature(const char *buffer, temperature_t *temperature, struct p
 static void sampletime(const char *buffer, duration_t *time)
 {
 	int i;
-	int hr, min, sec;
+	int hr, min, sec, ms;
 
-	i = sscanf(buffer, "%d:%d:%d", &hr, &min, &sec);
+	i = sscanf(buffer, "%d:%d.%d", &min, &sec, &ms);
```

Comment:
What is your reason for not leaving the support for (optional) hours in the version with millisecond support?

Reply:
Because I prefer to be strict and ignore what others maybe do, I only want to ingest what I created and leave the rest up to the previous path.

Reply:
And what benefit do you see in this approach?

Reply:
I try not to assume something I'm not sure of whether it even exists.
Do you have a real use case in mind? Some software that already uses this format?

Reply:
And what disadvantages do you see in implementing a parser that is accomodating to sensible future formats?

Reply:
YAGNI.

We can still do that once this future format pops up.

Reply:
I don't really see how 'YAGNI' should be a justification for adding unneeded complexity by supporting a format of 'hh:mm:ss OR mm:ss.xxx BUT NOT hh:mm:ss.xxx' instead of just 'hh:mm:ss.xxx'...

Reply:
> 'hh:mm:ss OR mm:ss.xxx BUT NOT hh:mm:ss.xxx' instead of just 'hh:mm:ss.xxx'...

Currently this enables `mm:ss.xxx` (as that's the one Subsurface uses when it writes the values) OR all the previous supported variations.

If I'm not mistaken that's all previous supported variations of a duration:

```
(-)mm
mm.ss -- that one's patched to mm:ss before invocation of `sampletime()`
(-)mm:ss
(-)hh:mm:ss
```

So you want that all those formats also get sub second support?

What should we do with `mm.ss`? Keep on patching it before or add support built-in?

Reply:
I'd treat `mm.ss` as a special case - it looks like this was added in specifically to fix inconsistent use of the format in one specific version of a third party app.
And for adding milliseconds to `mm` a reasonable approach to write this duration will be to use `mm:00.xxx`, yielding the `mm:ss.xxx` case - so all we are left with is `mm`, `mm:ss[.xxx]`, and `hh:mm:ss[.xxx]`, plus the special case.

Reply:
Fine like that?

Reply:
Yes, the format support looks good now, thank you. Not sure if there is a better way to do the parsing in C++, so this is potentially another case of waiting for #4198 to drop and then do it the C++ way - @bstoeger should know more.

Reply:
I'm not aware of a better way than `scanf` in the standard library.

I'm not following closely, but there appears to be a proposal for an `std::format` counterpart: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1729r1.html
However, that will not be available for many years, if ever.

Reply:
> I'm not aware of a better way than `scanf` in the standard library.

Could you please have a look at the latest version I implemented in `core/units.c`? That one doesn't use `scanf` anymore.

Reply:
Where do I see that?

Reply:
In this PR, the latest commit.

Reply:
I'm not against this code, but it seems a bit inconsistent to me with respect to whether or not the separator is consumed. It would be great if you could write some tests (e.g. in `tests/testhelper.cpp`?) that makes the expected outcome clear.